Sort by type

daniel-beck opened the pull request:
Add sorting/grouping by type, similar to how the deck lists on mtgtop8.com are organized:

1. Lands
1. Creatures
1. Spells
1. Other

Preference of group membership descending in that order, i.e. Dryad Arbor is with the other lands, not creatures.

Actual code changes in the first commit, update to the generated cards list JS file in the second in case that's undesired.

JS code change is a giant copy and paste / search and replace of the CMC sort variant.

Nightfirecat commented:
I meant to comment on this over the weekend, but it got away from me.

The code looks good as far as I can tell, though I haven't actually tested it or given it a thorough look yet. I'll try to do that tonight.

I don't know about @marumari's preference, but I'd prefer the card list update be in a separate PR (or just made by @marumari directly). Keeping separate update efforts in different PRs is easier to manage.

marumari commented:
Actually, I'll go update the generated cards now, so if you could just send the sorting by type commit, I'd appreciate that.  Thanks!

Nightfirecat commented:
Testing from 7f136a4 yielded the "Type" sort option working exactly like the "Alphabetically" sort option. Can't see any obvious bugs in the code, but it's just not working for me. (Seems to be working fine in 71ec592?)

In any event, even when it is working, doing a test entry for cards of all types seems to return an interesting (and not fully correct) list; with the following input:

```
1 verdant catacombs
1 lightning bolt
1 thoughtseize
1 scavenging ooze
1 liliana of the veil
1 mishra's bauble
1 phyrexian unlife
1 bitterblossom
```
I was given the following result:

> 1 Verdant Catacombs
>
> 1 Scavenging Ooze
>
> 1 Lightning Bolt
> 1 Thoughtseize
>
>1 Bitterblossom
> 1 Liliana of the Veil
> 1 Mishra's Bauble
> 1 Phyrexian Unlife

Nightfirecat commented on js/cards/parsecards.py:
```diff
@@ -72,6 +73,12 @@ def getLegalities(card, cards):
     elif cards[card]['colors'] == ['Red']:    ocards[ocard]['c'] = 'D'
     elif cards[card]['colors'] == ['Green']:  ocards[ocard]['c'] = 'E'
 
+    if   'Land'     in cards[card]['types']: ocards[ocard]['t'] = '1'
```
Can we get an extra space between the test case (`'Land'`) and the `in` keyword so that the type lines up with the the color type lineup above?

daniel-beck replied:
Adding an extra space after the colon as that's how it works in the list above.

Nightfirecat commented on js/decklist/decklist.js:
```diff
@@ -303,6 +307,56 @@ function sortDecklist(deck, sortorder) {
+      else { type = 'z'; } // Unknown
+
+      // Create the type subarray
+      if ( !(type in type_to_cards ) ) { type_to_cards[type] = []; }
```
Why the extra space in the test condition? `!(type in type_to_cards )`

daniel-beck replied:
@Nightfirecat 

For consistency with the rest of the file:

      if ( !(color in color_to_cards ) ) { color_to_cards[color] = []; }
      if ( !(cmc in cmc_to_cards ) ) { cmc_to_cards[cmc] = []; }
      if ( !(type in type_to_cards ) ) { type_to_cards[type] = []; }

Personally, this style makes my eyes bleed, but it's not my project, I'm just a guest here.

Nightfirecat replied:
Ah, ok. I hadn't noticed it above; carry on.

Nightfirecat commented on js/decklist/decklist.js:
```diff
@@ -303,6 +307,56 @@ function sortDecklist(deck, sortorder) {
+
+      // Fix the Aetherling issue until the PDF things supports it
+      lcard = lcard.replace('\u00c6', 'Ae').replace('\u00e6', 'ae');
+      deck[i][0] = deck[i][0].replace('\u00c6', 'Ae').replace('\u00e6', 'ae');
```
We really should just make a function for this... Not a nitpick for this PR, but I may take this up separately.

Nightfirecat commented on js/decklist/decklist.js:
```diff
@@ -303,6 +307,56 @@ function sortDecklist(deck, sortorder) {
+
+      type = type_to_cards_keys[i];
+
+      type_to_cards[ type ].sort();   // type_to_cards[3]
```
Comment in this line should be removed?

daniel-beck replied:
@Nightfirecat  Removing it.

daniel-beck commented:
@Nightfirecat 

> I'd kinda prefer if each type was separate; having enchantments and artifacts separate at least would be nice in my opinion, but I may be alone in that thought.

As I wrote in the PR description, I implemented the grouping used on mtgtop8.com, which seems reasonably popular.

daniel-beck commented:
@marumari @Nightfirecat 

Thanks for your review.

I removed the generated cardlist commit from my branch, and added a new commit addressing review comments. Other review comments have been addressed in comment responses to this PR.

Nightfirecat commented:
Thanks for the quick fix. LGTM :ship: :it:

marumari commented:
Someday I hope to go back and rewrite a lot of this code using lodash and make decks a bit more object-oriented, which will clean up a lot of stuff.  I just haven't had the time.  :)

daniel-beck commented:
@marumari @Nightfirecat Any chance to get this merged and deployed soon?

marumari commented:
So when I pulled it down and took a look at it with it sorting by type, it looks exactly the same as sorting by original.

file:///Source/decklist/index.html?deckmain=2%09Arid%20Mesa%0A4%09Brainstorm%0A1%09Council%27s%20Judgment%0A4%09Counterbalance%0A2%09Counterspell%0A2%09Entreat%20the%20Angels%0A4%09Flooded%20Strand%0A4%09Force%20of%20Will%0A4%09Island%0A3%09Jace%2C%20the%20Mind%20Sculptor%0A2%09Plains%0A2%09Polluted%20Delta%0A4%09Ponder%0A2%09Scalding%20Tarn%0A4%09Sensei%27s%20Divining%20Top%0A3%09Snapcaster%20Mage%0A4%09Swords%20to%20Plowshares%0A4%09Terminus%0A3%09Tundra%0A2%09Volcanic%20Island&deckside=1%09Council%27s%20Judgment%0A1%09Counterspell%0A1%09Disenchant%0A3%09Flusterstorm%0A1%09Keranos%2C%20God%20of%20Storms%0A1%09Path%20to%20Exile%0A1%09Pyroblast%0A2%09Red%20Elemental%20Blast%0A2%09Rest%20in%20Peace%0A1%09Supreme%20Verdict%0A1%09Surgical%20Extraction%0A

So maybe I'm doing something horribly wrong, but it doesn't seem to work for me.

daniel-beck commented:
@marumari You need to run `parsecards.py` to regenerate the file [you wanted me to drop from the PR](https://github.com/marumari/decklist/pull/27#issuecomment-239865695). I force-pushed over the first commit to remove it from the history of this PR.

marumari commented:
Okay!  I'll do that tonight and get this merged.  :)

marumari commented:
Thanks for this!  Sorry for the delay, life is crazy with a 4 year old and a 5 month old.  O_o

marumari commented:
Pushed to decklist.org.  :)

daniel-beck commented:
@marumari Thank you!

marumari commented:
And thank you!  I really appreciate the contribution!
